Fix part of #6240: Profile Photo Upload E2E Test

U8NWXD opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Test that a profile photo can be uploaded on the preferences page.

Fixes part of #6240.

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

U8NWXD committed:
Preferences E2E Test: Test Profile Photo Upload

Test that a profile photo can be uploaded on the preferences page.

Fixes part of #6240.

U8NWXD committed:
Protractor: Test Uploading Too Large Profile Photo

Create protractor e2e test to check that when uploading a file over 1
MB, an error message is displayed.

The image used comes from the National Archives at College Park via
Wikimedia:
https://commons.wikimedia.org/wiki/File:%22(The)_English_Kitchen,_meals_20_cts._Broadway_St.,_Round_Pond,_OkIa.%22_By_Kennett,_January_1894_-_NARA_-_540093.jpg
The image is in the public domain as a work of the US Federal
Government, according to Wikimedia.

U8NWXD committed:
Move Dummy Large Image to Unbreak Backend Tests

Move Dummy Large Image out of the `tests/data/dummy_assets` directory to
unbreak backend test
`utils_test.UtilsTests.test_get_exploration_components_from_dir_with_no_yaml_file`.

Fixes part of #6240.

Showtim3 commented on core/tests/protractor_utils/PreferencesPage.js:
```diff
@@ -41,6 +42,44 @@ var PreferencesPage = function() {
+  this.submitProfilePhoto = function(imgPath) {
+    return this.uploadProfilePhoto(imgPath).then(function() {
+      waitFor.visibilityOf(
+        profilePhotoCropper, 'photo cropper still invisible');
```
I suggest this should be `photo cropper taking too long to appear`

U8NWXD replied:
Thanks for the suggestion! Done in 266e70f.

Showtim3 commented on core/tests/protractor_utils/PreferencesPage.js:
```diff
@@ -41,6 +42,44 @@ var PreferencesPage = function() {
+      profilePhotoSubmitButton.click();
+    }).then(function() {
+      return waitFor.invisibilityOf(
+        profilePhotoUploadInput, 'photo upload still visible');
```
Same here! `Photo upload is taking too long to disappear`

U8NWXD replied:
Done in 266e70f

Showtim3 reviewed (commented):
This is a hard one, Uploading photos and all. So congrats for this awesome PR.

U8NWXD committed:
Make Timeout Error Messages Clearer

When we wait for an element to appear or disappear before continuing the
tests, we specify an error message for if the wait times out. This
commit makes those error messages more clear.

Fixes part of #6240. Addresses review comments from @Showtim3.

U8NWXD commented:
@nithusha21 PTAL

nithusha21 reviewed (approved):
I think this looks amazing! Thanks @U8NWXD!

nithusha21 commented:
@vojtechjelinek @kevinlee12 @DubeySandeep @Showtim3 PTAL! Thanks :)

kevinlee12 commented on core/tests/protractor_desktop/preferences.js:
```diff
@@ -26,6 +26,31 @@ describe('Preferences', function() {
+      );
+  });
+
+  it('shows an error if uploaded photo is too large', function() {
```
`should show and error...`

Always begin these tests with should or should not.

U8NWXD replied:
Fixed in 538c9e0

kevinlee12 commented on core/tests/protractor_desktop/preferences.js:
```diff
@@ -26,6 +26,31 @@ describe('Preferences', function() {
     preferencesPage = new PreferencesPage.PreferencesPage();
   });
 
+  it('should upload a profile photo', function() {
```
`should let a user upload a ....`

U8NWXD replied:
Done in 538c9e0

kevinlee12 reviewed (changes requested):
Hi @U8NWXD, left two comments. As for the image, is it possible to use something that we already have in the codebase (I'm wary of getting into licensing issues).

U8NWXD commented:
@kevinlee12 I share your concern about licensing, which is why I chose a public-domain image (more details in the commit message: 0eb1880). The uploader rejects images over 1M, but I haven't found an image in the code base that is that large:

```console
$ # Output is of the form: (image size in kilobytes)  (image path)
$ find -X . -iname "*.jpg" -print0 2>/dev/null | xargs -0 du -k | sort -nr | head -1
192	./third_party/static/ng-img-crop-0.3.2/test/test.jpg
$ find -X . -iname *.png -print0 2>/dev/null | xargs -0 du -k | sort -nr | head -1
460	./data/explorations/binary_search/assets/image/elimination_height_419_width_1596.png
$ find -X . -iname "*.svg" -print0 2>/dev/null | xargs -0 du -k | sort -nr | head -1
820	./third_party/static/fontawesome-free-5.9.0-web/webfonts/fa-solid-900.svg
$ find -X . -iname "*.gif" -print0 2>/dev/null | xargs -0 du -k | sort -nr | head -1
124	./node_modules/jasmine-spec-reporter/screenshot.gif
$ find -X . -iname "*.raw" -print0 2>/dev/null | xargs -0 du -k | sort -nr | head -1
```

All of these images are under the 1024k we would need.

kevinlee12 commented:
Ah, okay, let's leave the image as-is then. thanks!

DubeySandeep commented on core/tests/protractor_desktop/preferences.js:
```diff
@@ -26,6 +26,31 @@ describe('Preferences', function() {
+    expect(preferencesPage.getProfilePhotoSource())
+      .not
+      .toEqual(
+        preferencesPage.submitProfilePhoto('../data/img.png')
```
@U8NWXD I think this will not work in prod_env, have you tried running the test with prod_env flag? (#6687 will be helpful, maybe?)

U8NWXD replied:
@DubeySandeep this appears to work with the `prod_env` flag:

```console
$ bash scripts/run_e2e_tests.sh --prod_env --suite=preferences --sharding=false
...
Started
Jasmine started
.
  Preferences
    ✓ should upload a profile photo
.    ✓ shows an error if uploaded photo is too large
.    ✓ should change editor role email checkbox value
.    ✓ should change feedback message email checkbox value
.    ✓ should set and edit bio in user profile
.    ✓ should change prefered audio language of the learner
.    ✓ should change prefered site language of the learner
.    ✓ should load the correct dashboard according to selection




8 specs, 0 failures
...
```

DubeySandeep replied:
@U8NWXD, I tried running server locally in prod_env and tried uploading images, found that it fails with 500. Can you please check how it is passing e2e tests?

![image](https://user-images.githubusercontent.com/16653571/64486787-56adf080-d24f-11e9-85ab-2686973df558.png)

DubeySandeep replied:
Hi @U8NWXD, are you able to reproduce this issue? Let me know if you are facing any issue?

kevinlee12 replied:
@U8NWXD ping?

U8NWXD replied:
Sorry for the delayed response; I've been off the grid for the past week or so. I'll work on reproducing and debugging this issue

U8NWXD replied:
@DubeySandeep I was unable to reproduce this error. Here are the steps I followed:

* Run server locally with `python -m scripts.start`
* Select `Sign In` from the top menu bar
* Create an account with email `test@example.com` and username `test`
* Select `Create` from the top menu bar
* Select the introduction block for editing
* Choose to upload `oppia/core/tests/data/img.png`
* Set caption to `hi` and accessibility text to `foobar`
* Click `Use This Image`
* Click `Save`

No error occurred, and the image was added to the introduction block.

I was running after merging in `develop` such that the last non-merge commit is ec254cfcc87843c36e24d68d335cc811536fb2d4. I have updated this PR with this merge. If you are still seeing this error on the updated branch, could you send me steps to reproduce?

DubeySandeep replied:
@U8NWXD, All the steps you have mentioned are correct except step 1, make it "Run server locally with `python -m scripts.start --prod_env`" as the e2e test on Travis runs in the prod mode.

U8NWXD replied:
@DubeySandeep When running in prod mode on develop, I'm unable to create a user via the `Sign In` menu bar option. After choosing to use `test@example.com` as the email, I get a blank page at `http://localhost:8181/signup?return_url=%2F` instead of the page where I select a username and agree to the terms of service.

Based on the server output to the console, this appears to be because of a 404 when retrieving `third_party/generated/js/third_party.js`. The file that exists (and that I think should be the target) is `third_party/generated/js/third_party.min.js`.

Output:

```
...
INFO     2019-09-20 04:46:34,396 module.py:835] default: "GET /third_party/generated/css/third_party.css HTTP/1.1" 404 -
INFO     2019-09-20 04:46:34,465 module.py:835] default: "GET /templates/dev/head/css/oppia.css HTTP/1.1" 304 -
INFO     2019-09-20 04:46:34,606 module.py:835] default: "GET /webpack_bundles/signup.bundle.js HTTP/1.1" 304 -
INFO     2019-09-20 04:46:34,652 module.py:835] default: "GET /webpack_bundles/about%7Eadmin%7Eclassroom%7Ecollection_editor%7Ecollection_player%7Ecommunity_dashboard%7Econsole_errors%7Econtact%7Ee78591d4.bundle.js HTTP/1.1" 304 -
INFO     2019-09-20 04:46:34,746 module.py:835] default: "GET /webpack_bundles/runtime.bundle.js HTTP/1.1" 304 -
INFO     2019-09-20 04:46:35,703 module.py:835] default: "GET /webpack_bundles/vendors%7Eabout%7Eadmin%7Eclassroom%7Ecollection_editor%7Ecollection_player%7Ecommunity_dashboard%7Econsole_errors%7E561518cc.bundle.js HTTP/1.1" 304 -
INFO     2019-09-20 04:46:36,339 module.py:835] default: "GET /third_party/generated/js/third_party.js HTTP/1.1" 404 -
```

DubeySandeep replied:
@U8NWXD, Have you tried updating your branch with upstream/develop and restarting server? (though I don't expect this to happen for an outdated branch as well.)

U8NWXD replied:
@DubeySandeep Yes I have, but the problem persists, even after I ran
```
rm -rf ../oppia_tools/ node_modules/ third_party/ core/templates/prod/ local_compiled_js/
```
I'm now trying with a new clone of the oppia repo

U8NWXD replied:
@DubeySandeep I was able to reproduce your error, and I filed #7686 for it. That error occurs when uploading a photo to an exploration, but not when uploading a profile photo. My e2e test passes despite the error because I'm only testing uploading profile photos, not uploading exploration photos.

U8NWXD replied:
@DubeySandeep It seems to me that this bug is separate from (and not intended to be caught by) the e2e test introduced by this PR. Do you agree?

DubeySandeep replied:
> @DubeySandeep It seems to me that this bug is separate from (and not intended to be caught by) the e2e test introduced by this PR. Do you agree?

That's not a bug, its an expected behavior in prod_env we store images in GoogleCloudService but locally GCS key are not available so the upload fails.

I was not sure whether we upload profile images on GCS or in the app engine datastore itself. If we store images in AppEngine itself then I'm fine with this PR. (I'll check, make sure and approve this PR.)

DubeySandeep replied:
I confirm that we upload a base64 encoded profile image in the AppEngine itself. (This is only for the profile pictures, not for the other images such as images in exploration card.)

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

U8NWXD committed:
Update Naming of Profile Photo Upload E2E Tests

Make all test names of the form `should ...` and descriptive of what the
site should do.

Fixes part of #6240. Addresses review comments from @kevinlee12.

U8NWXD commented:
@kevinlee12 @DubeySandeep PTAL

kevinlee12 reviewed (approved):
lgtm!

U8NWXD committed:
Merge remote-tracking branch 'upstream/develop' into protractor-upload-profile-photo

U8NWXD committed:
Merge remote-tracking branch 'upstream/develop' into protractor-upload-profile-photo

U8NWXD commented:
Could someone please assign this PR to @DubeySandeep?

seanlip commented:
Hi @U8NWXD, done --thanks for the ping!

U8NWXD commented:
Frontend tests failed on CircleCI due to a timeout. They pass when I run them locally, both with and without the `--run_minified_tests --skip_install` arguments. Is this a known problem? Any suggestions on how I should proceed?

DubeySandeep commented:
@U8NWXD, I've restarted the test I think the will pass now.

Adding logs for @oppia/dev-workflow-team:  

```
#!/bin/bash -eo pipefail
python -m scripts.run_frontend_tests --run_minified_tests --skip_install
Checking if node.js is installed in /home/circleci/oppia/../oppia_tools
Checking if yarn is installed in /home/circleci/oppia/../oppia_tools
Environment setup completed.
Checking whether Google App Engine is installed in /home/circleci/oppia/../oppia_tools/google_appengine_1.9.67/google_appengine
Checking whether google-cloud-sdk is installed in /home/circleci/oppia/../oppia_tools/google-cloud-sdk-251.0.0/google-cloud-sdk
Running build task with concatenation and minification
Building third party libs at third_party/generated/
Compiling ts files...
Minifying and creating sourcemap for third_party/generated/js/third_party.js
Computing hashes for files in /home/circleci/oppia/assets/
Computing hashes for files in /home/circleci/oppia/extensions/
Computing hashes for files in /home/circleci/oppia/local_compiled_js/extensions/
Computing hashes for files in /home/circleci/oppia/core/templates/dev/head/
Computing hashes for files in /home/circleci/oppia/local_compiled_js/core/templates/dev/head/
Computing hashes for files in /home/circleci/oppia/third_party/generated/
Building webpack
Starting type checking service...
Using 1 worker with 2048MB memory limit
Too long with no output (exceeded 10m0s)
```

U8NWXD commented:
@DubeySandeep the frontend tests are still failing on CircleCI

DubeySandeep commented:
@U8NWXD, I've created an issue at #7646 and we are trying to look into this issue. (Sorry for the trouble, as of now you can concentrate on the e2e test PR.) 

/cc @oppia/dev-workflow-team

U8NWXD commented:
Please assign @DubeySandeep
